### PR TITLE
DAOS-4535 control: run server service as user

### PR DESCRIPTION
**Pull request by tanabarr**
When daos_server is installed from RPM, systemd starts service as daos
user and attempts to load certificates and config from /etc/daos/ by
default. Blank config is loaded on install and daos_server starts in
discovery mode with no configured IO servers.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by brianjmurrell** [comment]
Why is this PR in a draft state?

File: doc/admin/deployment.md [hunk 1]
```diff
@@ -365,36 +365,39 @@ The --enable-recovery is required for fault tolerance to guarantee that
...
 
-Once the service file is installed you can start daos_server
+Once the service file is installed you can start `daos_server`
 with the following commands:
 
 ```bash
-$ systemctl enable daos-server
-$ systemctl start daos-server
+$ systemctl enable daos_server
+$ systemctl start daos_server
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:388]: I believe it's most correct to put the type of unit you are working on on the unit name:

```suggestion
$ systemctl enable daos_server.service
$ systemctl start daos_server.service
```

File: doc/admin/deployment.md [hunk 2]
```diff
@@ -365,36 +365,39 @@ The --enable-recovery is required for fault tolerance to guarantee that
...
 ```
 
 To check the component status use:
 
 ```bash
-$ systemctl status daos-server
+$ systemctl status daos_server
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:394]:
```suggestion
$ systemctl status daos_server.service
```

File: doc/admin/deployment.md [hunk 3]
```diff
@@ -365,36 +365,39 @@ The --enable-recovery is required for fault tolerance to guarantee that
...
 ```
 
 If DAOS Server failed to start, check the logs with:
 
 ```bash
-$ journalctl --unit daos-server
+$ journalctl --unit daos_server
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:400]:
```suggestion
$ journalctl --unit daos_server.service
```

File: doc/admin/deployment.md [hunk 4]
```diff
@@ -894,37 +897,38 @@ $ daos_agent -i -o <'path to agent configuration file/daos_agent.yml'> &
 
 Alternatively, the DAOS Agent can be started as a systemd service. The DAOS Agent
 unit file is installed in the correct location when installing from RPMs.
+
 If you wish to use systemd with a development build, you must copy the service
-file from utils/systemd to /usr/lib/systemd/system. Once the file is copied
-modify the ExecStart line to point to your in tree daos_agent binary.
+file from utils/systemd to `/usr/lib/systemd/system`. Once the file is copied
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:902]:
```suggestion
file from `utils/systemd` to `/usr/lib/systemd/system`. Once the file is copied
```

File: doc/admin/deployment.md [hunk 5]
```diff
@@ -894,37 +897,38 @@ $ daos_agent -i -o <'path to agent configuration file/daos_agent.yml'> &
...
 
-Once the service file is installed, you can start daos_agent
+Once the service file is installed, you can start `daos_agent`
 with the following commands:
 
 ```bash
-$ sudo systemctl enable daos-agent
-$ sudo systemctl start daos-agent
+$ sudo systemctl enable daos_agent
+$ sudo systemctl start daos_agent
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:912]:
```suggestion
$ sudo systemctl enable daos_agent.service
$ sudo systemctl start daos_agent.service
```

File: doc/admin/deployment.md [hunk 6]
```diff
@@ -894,37 +897,38 @@ $ daos_agent -i -o <'path to agent configuration file/daos_agent.yml'> &
...
 ```
 
 To check the component status use:
 
 ```bash
-$ sudo systemctl status daos-agent
+$ sudo systemctl status daos_agent
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:918]:
```suggestion
$ sudo systemctl status daos_agent.service
```

File: doc/admin/deployment.md [hunk 7]
```diff
@@ -894,37 +897,38 @@ $ daos_agent -i -o <'path to agent configuration file/daos_agent.yml'> &
...
 ```
 
 If DAOS Agent failed to start check the logs with:
 
 ```bash
-$ sudo journalctl --unit daos-agent
+$ sudo journalctl --unit daos_agent
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:924]:
```suggestion
$ sudo journalctl --unit daos_agent.service
```

File: utils/rpms/daos.spec [hunk 8]
```diff
@@ -208,6 +208,7 @@ install -m 644 utils/systemd/%{agent_svc_name} %{?buildroot}/%{_unitdir}
 
 %pre server
 getent group daos_admins >/dev/null || groupadd -r daos_admins
+getent passwd daos_server >/dev/null || useradd -M -g daos_admins daos_server
```
Comment (brianjmurrell) [inline utils/rpms/daos.spec:211]: Or just `daos`?
```suggestion
getent passwd daos >/dev/null || useradd -M -g daos_admins daos
```

Maybe the group should also be just `daos` also.

File: utils/rpms/daos.spec [hunk 9]
```diff
@@ -257,6 +258,7 @@ getent group daos_admins >/dev/null || groupadd -r daos_admins
 
 %files server
 %config(noreplace) %{conf_dir}/daos_server.yml
+%attr(0660,root,daos_admins) %{conf_dir}/daos_server.yml
```
Comment (brianjmurrell) [inline utils/rpms/daos.spec:261]: Are there secrets in this file requiring the `660` permissions?

You are going to need to fix the _Test CentOS 7 RPMs" stage if so:

```
01:14:43.696  + cat /etc/daos/daos_server.yml
01:14:43.696  cat: /etc/daos/daos_server.yml: Permission denied
```

Comment (mjmac) [reply utils/rpms/daos.spec:261]: Yeah, I think this should be 0664 (world-readable). There's nothing secret in the server config.

File: utils/rpms/daos.spec (hunk 8 above)
Comment (mjmac) [reply utils/rpms/daos.spec:211]: The `daos_admins` group is used to control which processes are allowed to execute the `daos_admin` privileged helper. We set the `daos_server` binary to be setgid `daos_admins` and make `daos_admin` owned by `root,daos_admins` with permissions `0750`. I think it's worth keeping this separate from the `daos` user which is intended to be used for running the DAOS servers as an unprivileged user.

Comment (tanabarr) [reply utils/rpms/daos.spec:211]: won't "daos" user be potentially confused with any client application users with the same name in the case that the same host is used for client and server operations?

Comment (tanabarr) [reply utils/rpms/daos.spec]: never mind seems this is sufficiently low likelihood to not be a concern

Comment (brianjmurrell) [reply utils/rpms/daos.spec]: > won't "daos" user be potentially confused with any client application users with the same name in the case that the same host is used for client and server operations?

Are daos _client_ applications typically run in the context of and therefore as regular users?  That is, the client isn't a **service** on the machine but simply a binding of the client library to the end-user's process, which means it runs as that user'd ID/account?

So even on a server, if a user logs into the machine, as say, _tanabarr_ and runs an application that wants access to DAOS, it uses the daos client (i.e. libdaos) as _tanabarr_ doesn't it?

File: src/control/security/config.go [hunk 10]
```diff
@@ -33,9 +33,9 @@ import (
...
 const (
-	defaultCACert        = ".daos/daosCA.crt"
-	defaultServerCert    = ".daos/daos_server.crt"
-	defaultServerKey     = ".daos/daos_server.key"
+	defaultCACert        = "/etc/daos/daosCA.crt"
+	defaultServerCert    = "/etc/daos/server.crt"
+	defaultServerKey     = "/etc/daos/server.key"
 	defaultClientCert    = ".daos/client.crt"
 	defaultClientKey     = ".daos/client.key"
 	defaultAgentCert     = ".daos/agent.crt"
```
Comment (mjmac) [inline src/control/security/config.go:41]: Shouldn't the default locations for these certificates also be updated?

File: src/control/server/config.go [hunk 11]
```diff
@@ -394,6 +394,15 @@ func saveActiveConfig(log logging.Logger, config *Configuration) {
 
 // Validate asserts that config meets minimum requirements.
 func (c *Configuration) Validate(log logging.Logger) (err error) {
+	// config without servers is valid when initially discovering hardware
+	// prior to providing a fully populated configu
```
Comment (mjmac) [inline src/control/server/config.go:398]: typo: configu

File: src/control/server/config.go [hunk 12]
```diff
@@ -394,6 +394,15 @@ func saveActiveConfig(log logging.Logger, config *Configuration) {
...
+	if len(c.Servers) == 0 {
+		log.Infof("No %ss in configuration, %s starting in discovery mode",
+			DataPlaneName, ControlPlaneName)
+		c.Servers = nil
+		return nil
+	}
```
Comment (mjmac) [inline src/control/server/config.go:404]: Is this enough to allow the server to start up in discovery mode? I believe that there is another check for the fabric provider which will fail when it's blank.

Comment (tanabarr) [reply src/control/server/config.go:404]: it's working so I think it sufficient

File: src/control/security/config.go (hunk 10 above)
Comment (tanabarr) [reply src/control/security/config.go:41]: wasn't sure whether I should just do the server ones but after clarification from David , I will do all of them

File: src/control/server/config.go (hunk 12 above)
Comment (mjmac) [reply src/control/server/config.go]: Oh, right. I was thinking the fabric validation happened elsewhere, but it's right below this section.

**Comment by tanabarr** [comment]
CI intermittent failures in the rebuild tests:
DAOS-4357
DAOS-4630
DAOS-4626

**Comment by tanabarr** [comment]
LEAP15 optional check failed due to temporal issues as reported on https://status.opensuse.org/

File: doc/admin/deployment.md [hunk 13]
```diff
@@ -35,156 +35,6 @@ on boot if start-up scripts are registered with systemd.
...
-           /usr/share/spdk/scripts
-$ sudo ln -s $SL_PREFIX/include \
-           /usr/share/spdk/include
-```
-
-!!! note
-    The RPM installation is preferred for production scenarios. Manual
-    installation is most appropriate for development and predeployment
-    proof-of-concept scenarios.
-
```
Comment (brianjmurrell) [inline doc/admin/deployment.md:38]: Wow.  All of this is no longer needed?  That is awesome!

Comment (mjmac) [reply doc/admin/deployment.md:38]: If you install from RPM, that's correct. If you're running a developer build you still have to set things up manually. Those instructions just got moved to a new pre-flight document in beab74f.

**Pull request merged**